### DISCUSSION TITLE
Atmospheric constants

### DISCUSSION THREAD

**Discussion by gagamail** [imk-toolkit/imk-toolkit#47] [Q&A] [answered]
What should we do about some constants (Avogadro number, gas constant and so on) that are needed? 

1. Define them all in one file and use those definitions everywhere in the module
2. Define them locally in the functions

Will the first suggestion work or will such a file break some of the xarray checks?

**Comment by ucyo** [comment]
> Will the first suggestion work or will such a file break some of the xarray checks?

The xarray checks are independent of dependencies and technically it would work. If there are additional dependencies for the script, we will simply add them to the dependencies list of `imktk`. But... Do we want this?

> What should we do about some constants (Avogadro number, gas constant and so on) that are needed?

I think this is a bit early for shared variables across scripts. I can see several problems with this:
- People have different preferences with libraries providing these constants e.g. π and `math.pi` v `numpy.pi`
- What do we do if there are no units provided in the metadata
- The scripts would not be able to work anymore without `imktk`

The latter problem is I think the major issue. Currently the goal of the `imktk` module is to share scripts easily. This would add a whole other layer to it. I think support for SI units should rather be added to `xarray` or `numpy` itself and not in an external module. There are libraries out there which support unit operations coupled with `numpy` like [`si-units`](https://pypi.org/project/si-units/) or [`scimath`](https://docs.enthought.com/scimath/units/unit_numpy.html). This could be used for individual scripts of course, but I would not recommend using it on module level.

**Reply by t-redd** [reply]
I am strongly in favor to have a common definitions of constants in the toolkit. Local constants which may differ from file to file are unacceptable. Therefore I would suggest option 1 of Stefan's suggestion, together with the use of scipy.constants.

**Reply by msinn71** [reply]
I'm also not in favour of option 2 - defining constants locally in the functions could lead to inconsistencies between the functions, if different values are used.

**Reply by t-redd** [reply]
Just want to add to the points @ucyo raised:
- taste of constants: π is a nice example, I know mathematicians, they really put the digits into the code, and probably, using different values (differing in perhaps the last digit) is harmless (diff might be higher when defining it from a trigonometric function with lower precision, but one then would be consistent in the results for these functions at least). This we cannot solve and I prefer the consistent way.
- the question of units: CF formatted fields should have a unit attribute. I would silently assume SI units only. If not given, I would expose a warning.
- the questions of independence: this is still given for routines where you do not rely on consistent units, for example when calculating simple agregations and similar. But for example when converting to a number density the data input must not just be formally correct but needs to be a mixing ratio. Without further knowledge (eg. attributes) you cannot guarantee correct results. So you need a context anyway.

**Comment by ucyo** [comment, marked as answer]
If the majority is in favor of using common constants we could add this to the module. Please use emoticons to 👍🏼 or 👎🏼 this answer if you are in favor for/against adding this constants. 

*But please be aware, that if we choose to do it this way, it should be consistent in all current and future scripts. It will be enforced by the static analysis of the code and fail if constants are defined in functions.*

**Reply by ucyo** [reply]
CC @t-redd @msinn71 @gagamail

**Reply by t-redd** [reply]
Don't understand what you mean: the static analysis has no idea of the meaning of constants, I guess; if any constant in a function is forbidden, this is also not helpful, as it may be specific and used only there.

**Reply by ucyo** [reply]
@t-redd which use case are you thinking of? If a variable is a constant and we decide to go this way and want to make sure that everyone uses the same constants, the constants have to move to the constants module. That is the whole point in doing a constant module, right? No exceptions. If we wanna have exceptions, who decides which constants to use in the constant module and which not to? How would such a rule set look like?
